pass TypeMeta by value

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45263 use bitfield to shrink TensorImpl
* #45046 faster TensorOptions merging
* #45043 track Half/ComplexHalf default dtype
* **#45026 pass TypeMeta by value**

This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

Commit:
pass TypeMeta by value

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

smessmer reviewed (commented):
Have you done instruction count benchmarks? Given what we saw with `c10::optional`, size isn't all that matters in the by-value vs by-ref discussion, but triviality of copies/moves also matters.

Commit:
Update on "pass TypeMeta by value"


This is a big `s/TypeMeta&/TypeMeta/g` codemod. Before #44965 there was no size advantage to passing by value, but we would still have avoided an indirection. Now we get that and also a size reduction.

Differential Revision: [D23802943](https://our.internmc.facebook.com/intern/diff/D23802943)

[ghstack-poisoned]

bhosmer commented:
> Have you done instruction count benchmarks? Given what we saw with `c10::optional`, size isn't all that matters in the by-value vs by-ref discussion, but triviality of copies/moves also matters.

Yeah, worth calling out. On the empty()-from-C++ microbenchmark the instruction count is verrry slightly lower when we pass by value, which establishes that we're not inadvertently regressing.
